File: problems/HackerRank/2521_exercises/tree/kthSmallest.c
void dfs(struct TreeNode *node, int k, int *count, int *result)
{

	if (node == NULL || *count == k)
	{
		return;
	}

	dfs(node->left, k, count, result);

	(*count)++;

	if (*count == k)
	{
		*result = node->val;
	}

	dfs(node->right, k, count, result);
}

int kthSmallest(struct TreeNode *root, int k)
{

	int count = 0;
	int result = -1;

	dfs(root, k, &count, &result);
	return result;
}

File: problems/LeetCode/tree/kthSmallest.py
def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:

		sortedNodes = []

		def dfs(node: Optional[TreeNode]) -> None:
				if not node:
						return

				dfs(node.left)
				sortedNodes.append(node.val)
				dfs(node.right)
				
		dfs(root)

		return sortedNodes[k-1]


def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:

		def dfs(node: Optional[TreeNode]) -> int:

				nonlocal k
				if node == None:
						return None
				
				left_result = dfs(node.left)

				if left_result != None:
						return left_result

				k -= 1

				if k == 0:
						return node.val
				
				return dfs(node.right)
		
		return dfs(root)